source/Octopus.Tentacle.Tests/Diagnostics/Metrics/KubernetesAgentMetricsFixture.cs

```diff
@@ -0,0 +1,124 @@
+            items.First().Should().BeEquivalentTo(@event);
+        }
+
+        [Test]
+        public void CanWriteMultipleEntriesToThePersistenceMap()
+        {
+            MockPersistenceProvider persistenceProvider = new();
+            var sut = new KubernetesAgentMetrics(persistenceProvider, new MapFromConfigMapToEventList(), systemLog);
+
+            var events = new List<EventRecord>()
+            {
+                new EventRecord("Killed", "NFS Pod", DateTimeOffset.Now),
+                new EventRecord("Killed", "NFS Pod", DateTimeOffset.Now.AddMinutes(1))
+            };
+            
+            events.ForEach(e => sut.TrackEvent(e));
+
+            persistenceProvider.Content.Keys.Should().ContainSingle(entry => entry.Equals(KubernetesAgentMetrics.EntryName));
+            var items = JsonConvert.DeserializeObject<List<EventRecord>>(persistenceProvider.Content[KubernetesAgentMetrics.EntryName]);
+            items.Count.Should().Be(2);
+            items.Should().BeEquivalentTo(events);
+        }
+
+        [Test]
+        public void TrackEventDoesNotPropagateExceptions()
+        {
+            IPersistenceProvider persistenceProvider = Substitute.For<IPersistenceProvider>();
+            persistenceProvider.GetValue(Arg.Any<string>()).Throws(new Exception("Something broke"));
+            var sut = new KubernetesAgentMetrics(persistenceProvider, new MapFromConfigMapToEventList(), systemLog);
+            
+            var @event = new EventRecord("Killed", "NFS Pod", DateTimeOffset.Now);
+            Action act = () => sut.TrackEvent(@event);
+
+            act.Should().NotThrow();
+        }
+
+        [Test]
+        public void GetLatestTimestampReturnsDateTimeOffsetMinimumIfNoEventsExist()
+        {
+            MockPersistenceProvider persistenceProvider = new();
+            var sut = new KubernetesAgentMetrics(persistenceProvider, new MapFromConfigMapToEventList(), systemLog);
+
+            sut.GetLatestEventTimestamp().Should().Be(DateTimeOffset.MinValue);
+        }
+
+        [Test]
+        public void GetLatestTimestampReturnsTheChronologicallyLatestTimeNotNewestInList()
+        {
+            MockPersistenceProvider persistenceProvider = new();
+            var sut = new KubernetesAgentMetrics(persistenceProvider, new MapFromConfigMapToEventList(), systemLog);
+            
+            var events = new List<EventRecord>()
+            {
+                new EventRecord("Killed", "NFS Pod", DateTimeOffset.Now),
+                new EventRecord("Killed", "NFS Pod", DateTimeOffset.Now.AddMinutes(1)),
+                new EventRecord("Killed", "NFS Pod", DateTimeOffset.Now.AddMinutes(-1))
+            };
+
+            events.ForEach(e => sut.TrackEvent(e));
+
+            var timeDate = sut.GetLatestEventTimestamp();
+
+            timeDate.Should().BeExactly(events[1].Timestamp);
+        }
+
+        [Test]
+        public void GetLatestTimeStampPropagatesExceptionsIfUnderlyingPersistenceFails()
```

Comment:
What's the reason we want to propagate exceptions for this `GetLatestTimestamp` method but not the `TrackEvent` one?

Reply:
:thinking: we could propogate it up -and the layer above needs to catch it and handle the error.
I figured there was a risk that this would be called from critical path code, and if it fails, we didn't want to crash the calling function.

Whereas, the "GetValue" - if it doesn't work, we need to indicate failure, as subsequent steps may not be able to be conducted - potentially could do that by returning null - but that hides the fact that that something important down below failed.

I like the GetLatestEvent throwing - and can be convinced in having the TrackEvent throwing too.

There is a slight smell with the interface. -get TimeDate is a "hollow" - it probably should return the whole event at the very least ...